// yoomoney-sdk/index.js
const https = require('https');
const querystring = require('querystring');
const crypto = require('crypto');

/**
 * Minimal implementation of the yoomoney-sdk package.
 * Provides helper methods to interact with the YooMoney API using OAuth tokens.
 *
 * Supported endpoints:
 *  - account-info
 *  - operation-history
 *  - operation-details
 *  - request-payment
 *  - process-payment
 */
class API {
  constructor(accessToken, options = {}) {
    if (!accessToken) {
      throw new Error('YooMoney SDK: access token is required');
    }
    this.accessToken = accessToken;
    this.apiHost = options.apiHost || 'yoomoney.ru';
    this.timeout = options.timeout || 10000;
  }

  request(endpoint, params = {}) {
    return new Promise((resolve, reject) => {
      const postData = querystring.stringify(params);

      const req = https.request(
        {
          hostname: this.apiHost,
          path: `/api/${endpoint}`,
          method: 'POST',
          timeout: this.timeout,
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Content-Length': Buffer.byteLength(postData),
            Authorization: `Bearer ${this.accessToken}`
          }
        },
        (res) => {
          let data = '';
          res.on('data', (chunk) => {
            data += chunk;
          });
          res.on('end', () => {
            const body = data.trim();
            if (!body) {
              return resolve({});
            }
            try {
              resolve(JSON.parse(body));
            } catch (err) {
              try {
                resolve(querystring.parse(body));
              } catch (parseErr) {
                reject(parseErr);
              }
            }
          });
        }
      );

      req.on('timeout', () => {
        req.destroy(new Error('YooMoney SDK request timeout'));
      });

      req.on('error', (err) => {
        reject(err);
      });

      req.write(postData);
      req.end();
    });
  }

  accountInfo() {
    return this.request('account-info');
  }

  operationHistory(params = {}) {
    return this.request('operation-history', params);
  }

  operationDetails(params = {}) {
    return this.request('operation-details', params);
  }

  requestPayment(params = {}) {
    return this.request('request-payment', params);
  }

  processPayment(params = {}) {
    return this.request('process-payment', params);
  }
}

class YMPaymentFormBuilder {
  constructor(options = {}) {
    const {
      receiver,
      sum,
      label,
      successURL,
      targets,
      comment,
      quickpayForm = 'button',
      paymentType = 'AC',
      paymentTypeChoice = 'on'
    } = options;

    if (!receiver) {
      throw new Error('YMPaymentFormBuilder: receiver is required');
    }
    if (!sum) {
      throw new Error('YMPaymentFormBuilder: sum is required');
    }

    this.fields = {
      receiver,
      'quickpay-form': quickpayForm,
      'paymentType': paymentType,
      'paymentType-choice': paymentTypeChoice,
      sum: Number(sum).toFixed(2),
      label: label || '',
      successURL: successURL || '',
      targets: targets || 'Пополнение баланса',
      comment: comment || ''
    };
  }

  buildFormHTML() {
    return `
      <form method="POST" action="https://yoomoney.ru/quickpay/confirm" accept-charset="utf-8">
        ${Object.entries(this.fields)
          .filter(([, value]) => value !== undefined && value !== null && value !== '')
          .map(
            ([key, value]) =>
              `<input type="hidden" name="${key}" value="${String(value).replace(/"/g, '&quot;')}" />`
          )
          .join('\n')}
        <button type="submit" style="padding: 12px 20px; border-radius: 10px; border: none; background: #6c5ce7; color: #fff; font-size: 16px; cursor: pointer;">
          Оплатить через YooMoney
        </button>
      </form>
    `;
  }

  buildHtml(fullPage = false) {
    const formHTML = this.buildFormHTML();
    if (!fullPage) {
      return formHTML;
    }

    return `<!DOCTYPE html>
<html lang="ru">
  <head>
    <meta charset="UTF-8" />
    <title>Оплата через YooMoney</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <style>
      body {
        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif;
        background: #05060a;
        color: #fff;
        min-height: 100vh;
        display: flex;
        align-items: center;
        justify-content: center;
        padding: 24px;
      }
      .card {
        background: rgba(255, 255, 255, 0.05);
        border-radius: 16px;
        padding: 32px;
        width: min(480px, 100%);
        box-shadow: 0 20px 60px rgba(0,0,0,0.4);
        text-align: center;
      }
      h1 {
        font-size: 1.6rem;
        margin-bottom: 8px;
      }
      p {
        color: rgba(255,255,255,0.75);
        margin-bottom: 24px;
      }
      form {
        margin-top: 20px;
      }
    </style>
  </head>
  <body>
    <div class="card">
      <h1>Перевод через YooMoney</h1>
      <p>Нажмите кнопку ниже, чтобы открыть защищённую форму оплаты YooMoney.</p>
      ${formHTML}
    </div>
  </body>
</html>`;
  }
}

class YMNotificationError extends Error {
  constructor(message) {
    super(message);
    this.name = 'YMNotificationError';
  }
}

class YMNotificationChecker {
  constructor(secret) {
    if (!secret) {
      throw new Error('YMNotificationChecker: secret is required');
    }
    this.secret = secret;
    this.memoSet = new Set();
  }

  buildSignaturePayload(body) {
    const fields = [
      body.notification_type || '',
      body.operation_id || '',
      body.amount || '',
      body.currency || '',
      body.datetime || '',
      body.sender || '',
      body.codepro || '',
      this.secret,
      body.label || ''
    ];
    return fields.join('&');
  }

  verify(body) {
    const signatureString = this.buildSignaturePayload(body);
    const expectedHash = crypto.createHash('sha1').update(signatureString, 'utf8').digest('hex');
    const incomingHash = String(body.sha1_hash || '').toLowerCase();
    if (incomingHash !== expectedHash) {
      throw new YMNotificationError('Invalid YooMoney notification signature');
    }
  }

  middleware(options = {}, handler) {
    const memoEnabled = options.memo !== false;
    return (req, res, next) => {
      const payload = req.body || {};
      try {
        this.verify(payload);
        if (memoEnabled) {
          const key = payload.notification_id || payload.operation_id;
          if (key) {
            if (this.memoSet.has(key)) {
              throw new YMNotificationError('Duplicate YooMoney notification');
            }
            this.memoSet.add(key);
          }
        }
        return handler(req, res, next);
      } catch (error) {
        return next(error);
      }
    };
  }
}

module.exports = {
  API,
  YooMoney: API,
  YMPaymentFormBuilder,
  YMNotificationChecker,
  YMNotificationError
};
