// yoomoney-sdk/index.js
const https = require('https');
const querystring = require('querystring');

/**
 * Minimal implementation of the yoomoney-sdk package.
 * Provides helper methods to interact with the YooMoney API using OAuth tokens.
 *
 * Supported endpoints:
 *  - account-info
 *  - operation-history
 *  - operation-details
 *  - request-payment
 *  - process-payment
 */
class API {
  constructor(accessToken, options = {}) {
    if (!accessToken) {
      throw new Error('YooMoney SDK: access token is required');
    }
    this.accessToken = accessToken;
    this.apiHost = options.apiHost || 'yoomoney.ru';
    this.timeout = options.timeout || 10000;
  }

  request(endpoint, params = {}) {
    return new Promise((resolve, reject) => {
      const postData = querystring.stringify(params);

      const req = https.request(
        {
          hostname: this.apiHost,
          path: `/api/${endpoint}`,
          method: 'POST',
          timeout: this.timeout,
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Content-Length': Buffer.byteLength(postData),
            Authorization: `Bearer ${this.accessToken}`
          }
        },
        (res) => {
          let data = '';
          res.on('data', (chunk) => {
            data += chunk;
          });
          res.on('end', () => {
            const body = data.trim();
            if (!body) {
              return resolve({});
            }
            try {
              resolve(JSON.parse(body));
            } catch (err) {
              try {
                resolve(querystring.parse(body));
              } catch (parseErr) {
                reject(parseErr);
              }
            }
          });
        }
      );

      req.on('timeout', () => {
        req.destroy(new Error('YooMoney SDK request timeout'));
      });

      req.on('error', (err) => {
        reject(err);
      });

      req.write(postData);
      req.end();
    });
  }

  accountInfo() {
    return this.request('account-info');
  }

  operationHistory(params = {}) {
    return this.request('operation-history', params);
  }

  operationDetails(params = {}) {
    return this.request('operation-details', params);
  }

  requestPayment(params = {}) {
    return this.request('request-payment', params);
  }

  processPayment(params = {}) {
    return this.request('process-payment', params);
  }
}

module.exports = {
  API,
  YooMoney: API
};
